wordpress-mobile/WordPress-iOS
# Notifications: Sharing and Reminders

This PR adds two 'experimental' notifications features, that probably need some design review before they're ready to ship to everybody. Although as somebody whose daily work revolves a lot around notifications, I feel they might be useful to ship in internal builds straight away to gather feedback. 

Both features are accessible via a new `...` "More" item, visible if you swipe left-to-right on a notification:

<img width="375" alt="Screenshot 2019-04-29 at 15 43 58" src="https://user-images.githubusercontent.com/4780/56904209-c0461300-6a95-11e9-9dda-6a28724e790a.png">

This is one thing I'd like feedback on – I'm not sure it's the best UI, but I don't know what would be better. Perhaps a long press? 

Select the `...` item to see options for each of the features:

<img width="373" alt="Screenshot 2019-04-29 at 15 46 04" src="https://user-images.githubusercontent.com/4780/56904305-ea97d080-6a95-11e9-9716-ef5a9abef9a3.png">

---

### Share

Share launches the system share sheet, sharing the notification's URL and subject. This is useful for sending a link to others, or storing in your todo app of choice:

![note-shares](https://user-images.githubusercontent.com/4780/56903824-f9ca4e80-6a94-11e9-9ec9-167bb690510b.gif)

--- 

### Reminders

The other option, Reminders, presents a list of time periods after which you'll be notified again about the notification. I used the same periods that Slack uses: 20 minutes, 1 hour, 3 hours, tomorrow (9am), next week (Monday, 9am).

![note-reminds](https://user-images.githubusercontent.com/4780/56903835-fd5dd580-6a94-11e9-885a-bab278349d6e.gif)

The notifications are iOS local notifications, so everything is specific to the device – you won't get a notification on a different device, or on the web. If you tap the system notification, you'll be taken to the detail view for the notification you've been reminded about.

When you have reminders scheduled, you can access them via a reminders icon in the top of the Notifications screen which only appears when you have some scheduled. This presents a list of all scheduled reminders. You can swipe them to cancel, or cancel them all with a single tap.

![reminders-1](https://user-images.githubusercontent.com/4780/56904861-0fd90e80-6a97-11e9-87f0-6139d7d268f3.png)

---

Let me know what you think! Is there anything that needs changing? Could we ship this internally for now?

**To test:**

* Build and run
* I added a few tests, so check those pass
* Try sharing some notifications
* Try setting reminders for some notifications and check they're triggered. If you want to make them trigger sooner than normal, you can alter `dateComponents(from date: Date)` in `NotificationReminderPeriod` in `NotificationRemindersHelper.swift`.

I haven't updated release notes yet, as this is still a draft and the functionality may change.

Update release notes:
